### PR TITLE
Replace BSQL with rust-g, paramaterize all queries

### PR DESCRIPTION
**Pull request by SpaceManiac**
:cl:
refactor: BSQL has been replaced with rust-g, improving SQL reliability and safety.
/:cl:

Queries were parameterized as much as possible. Careful code read by someone else also versed in SQL would be a good idea. Things seemed fine spot-testing roundstart, library, bans, and polls, but more stress-testing may be useful.

Also updates rust-g to v0.4.5:
- Improved Linux `.so` detection
- Perlin noise functions

Also updates Travis script to bionic.

### CODE REVIEW

**Comment by optimumtact** [comment]
@Jordie0608 pls

**Comment by Akrilla** [comment]
fucking :eyes:

**Comment by Cyberboss** [comment]
Most based PR of the year so far

**Comment by LemonInTheDark** [comment]
Absolute madman

**Comment by Archanial** [comment]
obraz
man versus machine

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 language: generic
 os: linux
-dist: xenial
+dist: bionic
```
Comment (Iamgoofball) [inline .travis.yml:3]: why not gentoo?

File: code/modules/admin/permissionedit.dm [hunk 2]
```diff
@@ -43,8 +36,14 @@
 				logcount -= logssperpage
 				pagecount++
 			output += "|"
-		var/limit = " LIMIT [logssperpage * page], [logssperpage]"
-		var/datum/DBQuery/query_search_admin_logs = SSdbcore.NewQuery("SELECT datetime, round_id, IFNULL((SELECT byond_key FROM [format_table_name("player")] WHERE ckey = adminckey), adminckey), operation, IF(ckey IS NULL, target, byond_key), log FROM [format_table_name("admin_log")] LEFT JOIN [format_table_name("player")] ON target = ckey[search] ORDER BY datetime DESC[limit]")
+		var/datum/DBQuery/query_search_admin_logs = SSdbcore.NewQuery({"
```
Comment (Cyberboss) [inline code/modules/admin/permissionedit.dm:39]: My bookmark, looks good above here

File: code/controllers/subsystem/dbcore.dm [hunk 3]
```diff
@@ -215,80 +216,94 @@ Delayed insert mode was removed in mysql 7 and only works with MyISAM type table
...
+				query_parts += ", "
+			if (has_question_mark[column])
+				var/name = "p[arguments.len]"
+				query_parts += replacetext(columns[column], "?", ":[name]")
+				arguments[name] = row[column]
+			else
+				query_parts += columns[column]
+			has_col = TRUE
+		query_parts += ")"
+		has_row = TRUE
```
Comment (MrStonedOne) [inline code/controllers/subsystem/dbcore.dm:262]: The messyness of this function perfectly sums up my argument against parameterized queries.

Comment (SpaceManiac) [reply code/controllers/subsystem/dbcore.dm:262]: It was messy before, too.

One alternative would be building a query with a single `VALUES` line and shipping multiple rows to rust-g to do a batch execute.

**Review by Jordie0608 — CHANGES_REQUESTED** [review]
wew

File: code/controllers/subsystem/blackbox.dm [hunk 4]
```diff
@@ -271,20 +286,13 @@ Versioning
...
-	sender = sender ? "'[sanitizeSQL(sender)]'" : "NULL"
-	var/server_ip = sanitizeSQL(world.internet_address)
-	var/server_port = sanitizeSQL(world.port)
-	var/round_id = sanitizeSQL(GLOB.round_id)
-
-	var/datum/DBQuery/query_log_ahelp = SSdbcore.NewQuery("INSERT INTO [format_table_name("ticket")] (ticket, action, message, recipient, sender, server_ip, server_port, round_id, timestamp) VALUES ('[ticket]', '[action]', '[message]', [recipient], [sender], INET_ATON(IF('[server_ip]' LIKE '', '0', '[server_ip]')), '[server_port]','[round_id]', '[SQLtime()]')")
+	var/datum/DBQuery/query_log_ahelp = SSdbcore.NewQuery({"
+		INSERT INTO [format_table_name("ticket")] (ticket, action, message, recipient, sender, server_ip, server_port, round_id, timestamp)
+		VALUES (:ticket, :action, :message, :recipient, :sender, INET_ATON(:server_ip), :server_port, :round_id, :time)
+	"}, list("ticket" = ticket, "action" = action, "message" = message, "recipient" = recipient, "sender" = sender, "server_ip" = world.internet_address || "0", "server_port" = world.port, "round_id" = GLOB.round_id, "time" = SQLtime()))
```
[the remark this answers is not in the document]
Comment (SpaceManiac) [reply code/controllers/subsystem/blackbox.dm:295]: Yes, this is handled. DM `null` becomes JSON `null` becomes SQL `NULL`.

File: code/controllers/subsystem/dbcore.dm [hunk 5]
```diff
@@ -215,80 +216,94 @@ Delayed insert mode was removed in mysql 7 and only works with MyISAM type table
...
-			var/idx = columns[column]
-			if (!idx)
-				idx = columns.len + 1
-				columns[column] = idx
-				sorted_row.len = columns.len
+			columns[column] = "?"
+			has_question_mark[column] = TRUE
+	for (var/column in special_columns)
+		columns[column] = special_columns[column]
+		has_question_mark[column] = !!findtext(special_columns[column], "?")
```
[the remark this answers is not in the document]
Comment (SpaceManiac) [reply code/controllers/subsystem/dbcore.dm:232]: Yes, it was unnecessary cleverness. Addressed.

File: code/controllers/subsystem/blackbox.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jordie0608) [reply code/controllers/subsystem/blackbox.dm:295]: All good then.

**Review by Jordie0608 — DISMISSED** [review]

**Comment by Cyberboss** [comment]
Time to break some servers

**Comment by Cyberboss** [comment]
I assume this will error if hardcoded VALUES are attempted?

**Comment by Cyberboss** [comment]
Also this is failing to test merge, can you pull master please?

**Comment by SpaceManiac** [comment]
> I assume this will error if hardcoded VALUES are attempted?

Not sure what you mean.

> Also this is failing to test merge, can you pull master please?

Done, but the conflict is probably with another testmerge

**Comment by Cyberboss** [comment]
To rephrase will the lib reject a query in the form 
```dm
"INSERT INTO SS13_round (id) VALUES ([round_id_var])"
```
String interpolated, not parameterized.

**Comment by SpaceManiac** [comment]
> will the lib reject a query

No, it can't tell the difference by that point

**Comment by Cyberboss** [comment]
I am still confused as to why parameterization is needed then. I can only
assume it's because there's no `sanitizeSQL` equivalent.

**Comment by SpaceManiac** [comment]
> I can only assume it's because there's no `sanitizeSQL` equivalent.

That's correct.

File: code/__HELPERS/text.dm [hunk 6]
```diff
@@ -13,10 +13,6 @@
  * SQL sanitization
  */
 
-// Run all strings to be used in an SQL query through this proc first to properly escape out injection attempts.
-/proc/sanitizeSQL(t)
-	return SSdbcore.Quote("[t]")
-
```
Comment (MrStonedOne) [inline code/__HELPERS/text.dm:19]:
```dm
/proc/quoteSQL(t)
	var/datum/DBQuery/query_escape_string = SSdbcore.NewQuery("SELECT QUOTE(:t)", list("t" = t))
	query_escape_string.Execute()
	. = query_escape_string.NextRow()
	qdel(query_escape_string)
```

:stuck_out_tongue:

Comment (optimumtact) [reply code/__HELPERS/text.dm:19]: you need help

**Comment by Cyberboss** [comment]
image

**Comment by spookydonut** [comment]
MSO: I don't like this
oranges: hahaha merge button goes brrrr

**Comment by optimumtact** [comment]
MSO has his reasons for preferring one form or the other, but in reality it doesn't make much of a difference to the size of the code.

People can write in either form with a modicum of effort.

What's more important is having a library that doesn't segfault our server every time the db server decides to take a 30 second nap.

plus it's easier to package on linux now as well.

**Comment by optimumtact** [comment]
That said, I do expect something to break from this

**Comment by optimumtact** [comment]
first breakage, compile error due to server side modifications

**Pull request merged**